docs/internals.rst

```diff
@@ -0,0 +1,332 @@
+
+Before publishing to the npm registry, `Babel <https://babeljs.io/>`__ compiles the code in the ``src`` directory and produces widely compatible ES5 JavaScript code in the ``lib`` directory. This is done because in the future we want to be able to run Dredd in the browser. With time this seems like a long shot and until we actually work on having Dredd browser-compatible, we might want to get rid of the build step.
+
+Tests need to be pre-compiled every time, because some integration tests use code linked from ``lib``. This is certainly a flaw and it slows down day-to-day development, but until we refactor the tests, compiling is necessary.
+
+Previously Dredd was written in `CoffeeScript <https://coffeescript.org>`__, and it was only recently converted to modern JavaScript. That's why sometimes the code does not feel very nice. Any efforts to refactor the code to something more human-friendly are greatly appreciated.
+
+CoffeeScript is still a production dependency (not dev dependency), because it’s needed for running user-provided hooks written in CoffeeScript. This is planned to be generalized: `#1082 <https://github.com/apiaryio/dredd/pull/1082>`__
+
+
+.. _cpp-dependencies:
+.. _compiled-vs-pure-javascript:
+
+C++ dependencies
+~~~~~~~~~~~~~~~~
+
+Dredd uses `Drafter <https://github.com/apiaryio/drafter>`__ for parsing `API Blueprint <https://apiblueprint.org/>`__ documents. Drafter is written in C++11 and needs to be compiled during installation. Because that can cause a lot of problems in some environments, there’s also pure JavaScript version of the parser, `drafter.js <https://github.com/apiaryio/drafter.js>`__. Drafter.js is fully equivalent, but it can have slower performance. Therefore there’s `drafter-npm <https://github.com/apiaryio/drafter-npm/>`__ package, which tries to compile the C++11 version of the parser and in case of failure it falls back to the JavaScript equivalent. Dredd depends on the `drafter-npm <https://github.com/apiaryio/drafter-npm/>`__ package.
+
+That’s the reason why even if you see ``node-gyp`` errors and failures during the installation process, afterwards Dredd seems to normally work and correctly parses API Blueprint documents.
+
+
+Avoiding C++ compilation
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``--no-optional`` option forces the JavaScript version of Drafter and avoids any compilation attempts when installing Dredd:
+
+.. code-block:: shell
+
+   $ npm install dredd --global --no-optional
+
+
+Troubleshooting C++ compilation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you need the performance of the C++11 parser, but you are struggling to get it installed, it’s usually because of the following problems:
+
+-  **Your machine is missing a C++11 compiler.** See how to fix this on `Windows <https://github.com/apiaryio/drafter/wiki/Building-on-Windows>`__ or `Travis CI <https://github.com/apiaryio/protagonist/blob/master/.travis.yml>`__.
```

Comment:
Supported compiler list is found at https://github.com/apiaryio/drafter/#compiler-support, we require more than C++11 these days.